### ISSUE TITLE
Remove `Constant.h/.cpp`

### ISSUE THREAD

**Issue by michalhabera** [FEniCS/dolfinx#175]
I'll continue my remove mood here.

I propose to remove `Constant` class as a part of dolfin. Let me explain:
`Constant` is used basically for 2 types of action:
1. In UFL expression (for assembly)
  In CPP workflow one works anyway with pure UFL constructs. FFCx then compiles it into e.g. `w[0][0]`, i.e. fixed index in array of coefficients.
  In Python workflow there is the ugly inheritance, because dolfin's constant is in fact ufl.Coefficient of "Real" finite element. ~Let alone the memory use for dolfin's Constant when allocating new Real element function...~ Ok, just one number, not really a memory deal.
  ~For both workflows the Constant is not really Constant. It is just another Coefficient.~ This is UFL thing.
2. As child of Expression elsewhere. In `DirichletBC` in order to fit to constructor, or in interpolation, ...
  Here, as it is just a type of Expression, it has to fit the generic interface and many optimisation possibilities are buried.

This task is also towards a stronger UFL/DOLFIN separation. See https://github.com/FEniCS/dolfinx/issues/150

I propose in steps:
1. Remove `Constant.h`, `Constant.cpp`
2. Specialise DirichletBC constructor for constant values (PetscScalar/std::vector<PetscScalar>/...)
  Basically make sure, that ufl.Constant and the constant valued DirichletBC are two **separate** concepts.
3. Create dispatch in DirichletBC before complicated `->restrict()` calls and follow to simpler constant codepath
4. For assembly the ufc signature should not change. Only preparation of "coefficient" vector should be simplified. I must have hands-on this to see what exactly could be a challenge.

I am willing to do the work. Comments please...

**Comment by michalhabera** [comment]
There is also the aspect of `Constant` in DirichletBC which changes value in "time" for every assembly step. I think it is better to call cirectly `bc.set_value(time...)` insted of changing some underlying Constant...

**Comment by garth-wells** [comment]
Another view is that maybe `Constant` should be improved to really be constant (over a spatial domain) in forms?

`Constant` is handy to make changes in forms without triggering jit, and without the overhead of a full `Expression`.

I agree that the flow in `DirichletBC` could be greatly simplified, and ideally the `restrict` call avoided in many cases

**Comment by michalhabera** [comment]
Ah, I see that deep in UFL, `Constant` is really a `Real` finite element coefficient cell-wise. Be it.

Two main points of my issue are 
1. #150 - do not wrap Constant in our python layer, prepare form on pure UFL language and attach problem specific info at compile time
2. Do not use Constant in DirichletBC. DirichletBC for constant value is too specific to dispatch to GenericFunction.

And for these two points no `dolfin.function.Constant` is needed.

Ad JIT: In `assemble()` methods, when problem specific data are attached, JIT should be able to cache.

Another strong argument is, that dolfin.function.Constant is ufl.Coefficient in python layer, but dolfin::function::Constant is GenericFunction in cpp layer. Very very confusing.

**Comment by michalhabera** [comment]
This will become clear with https://github.com/FEniCS/dolfinx/issues/183
